### PR TITLE
router: API cleanup for unit tests

### PR DESCRIPTION
**Pull request by alyssawilk**
Cleaning up APIs between router and upstream, and upstream and connection pools for testability.

Risk Level: Medium (router refactor) 
Testing: n/a
Docs Changes: n/a
Release N otes: n/a
Part of #1630 #1451

### CODE REVIEW

**Comment by alyssawilk** [comment]
Turns out the last set of refactors was enough to get everything working end to end, but testing the new TCP classes required creating both an UpstreamRequest and the full router filter and router config, which makes it kind of hard to unit test.

File: source/common/router/router.h [hunk 1]
```diff
@@ -377,14 +419,41 @@ class Filter : Logger::Loggable<Logger::Id::router>,
     return cookie_value;
   }
 
+  void onUpstream100ContinueHeaders(Http::ResponseHeaderMapPtr&& headers,
```
Comment (snowp) [inline source/common/router/router.h]: `// RouterFilterInterface` comment over these.

Do we care that the order of these in the .cc no longer follow the order in the .h file?

Comment (alyssawilk) [reply source/common/router/router.h:422]: I think preserving git history is more useful than having them in order, but happy to move them if others prefer the inverse.

Comment (snowp) [reply source/common/router/router.h:422]: That's a good enough reason for me

**Comment by alyssawilk** [comment]
wedged? 
/azp run

**Pull request merged**